{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###This is the README for all information downloadable from orthoDB\n",
    "Ortho DB v9 data dump consists of 9 files:\n",
    "\n",
    "odb9_levels.tab.gz:      NCBI taxonomy nodes where Ortho DB orthologous groups (OGs) are calculated\n",
    "odb9_species.tab.gz:     NCBI taxonomy nodes (mostly species level) participating in OGs\n",
    "odb9_genes.tab.gz:       genes with some info\n",
    "odb9_OGs.tab.gz:         Ortho DB orthologous groups\n",
    "odb9_OG2genes.tab.gz:    OGs to genes correspondence\n",
    "odb9v1_OG_xrefs.tab.gz:  OGs association with some external references, e.g. GO, COG and InterPro\n",
    "odb_v8_v9_1.tab          mappings between v8 and v9.1 orthologous groups\n",
    "\n",
    "odb9_fasta_<root>.tgz    tar-ball with one fasta file per taxon id in the given root (bacteria,metazoa,fungi,plants)\n",
    "\n",
    "\n",
    "The non-fasta files are in tab-separated format without column headers.\n",
    "\n",
    "The fasta files have headers with orthodb internal gene id as well as a public id.\n",
    "\n",
    "-----------------------------------------------------------------\n",
    "\n",
    "odb9_levels.tab:\n",
    "1.\tNCBI tax_id\n",
    "2.\tscientific name\n",
    "3.\ttotal non-redundant count of genes in all underneath clustered species\n",
    "4.\ttotal count of OGs built on it\n",
    "5.\ttotal non-redundant count of species underneath\n",
    "\n",
    "odb9_species.tab\n",
    "1.\tNCBI tax_id\n",
    "2.\tscientific name\n",
    "3.\ttotal count of clustered genes in this species\n",
    "4.\ttotal count of OGs it participates\n",
    "5.\tmapping type, clustered(C) or mapped(M)\n",
    "\n",
    "odb9_genes.tab\n",
    "1.\tOrtho DB unique gene id (not stable between releases)\n",
    "2.\torganism tax id\n",
    "3.\tprotein original sequence id, as downloaded together with the sequence\n",
    "4.\tUniprot id, evaluated by mapping\n",
    "5.\tENSEMBL gene name, evaluated by mapping\n",
    "6.\tNCBI gid, evaluated by mapping\n",
    "7.\tdescription, evaluated by mapping\n",
    "\n",
    "odb9_OGs.tab\n",
    "1.\tOG unique id (not stable between releases)\n",
    "2.\tlevel tax_id on which the cluster was built\n",
    "3.\tOG name (the group's most common gene name)\n",
    "\n",
    "odb9_OG2genes.tab\n",
    "1.\tOG unique id\n",
    "2.\tOrtho DB unique gene id\n",
    "\n",
    "odb9v1_OG_xrefs.tab\n",
    "1.\tOG unique id\n",
    "2.\texternal DB or DB section\n",
    "3.\texternal identifier\n",
    "4.\tnumber of genes in the OG associated with the identifier\n",
    "\n",
    "odb_v8_v9_1.tab\n",
    "1.  level tax_id on which both v8 and v9.1 OGs were built\n",
    "2.  v8   OG id\n",
    "3.  v9.1 OG id\n",
    "4.  distance between the two OGs (0. - identical, 1.0 - totally unrelated)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## This code allows you to query a Uniprot ID and will return a odb ID\n",
    "\n",
    "#!/usr/bin/env python3\n",
    "\n",
    "import sys\n",
    "\n",
    "query = sys.argv[1]\n",
    "ID_dict = dict()\n",
    "\n",
    "with open('odb9v1_genes.tab', 'r') as file_obj:\n",
    "    for line in file_obj:\n",
    "        line_list = line.split()\n",
    "        ID_dict[line_list[3]] = line_list[0]\n",
    "\n",
    "    print(ID_dict[query])\n",
    "## input of O14802    \n",
    "## returns 9606:00232f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## This code will return a OG from a provided odbID \n",
    "OG_dict = dict() \n",
    "\n",
    "with open('odb9v1_OG2genes.tab', 'r') as file_obj:\n",
    "    for line in file_obj:\n",
    "       OG_dict[line.split()[1]] = line.split()[0]\n",
    "    \n",
    "## input of 9606:00232f\n",
    "## returns EOG093700ID\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## This code will return a taxonomy ID for a OG provided \n",
    "\n",
    "tax_dict = dict()\n",
    "\n",
    "with open('odb9v1_OGs.tab', 'r') as file_obj:\n",
    "    for line in file_obj:\n",
    "        tax_dict[line.split()[0]] = line.split()[1]\n",
    "        \n",
    "## returns 2759"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "import sys\n",
    "\n",
    "query_file_name = sys.argv[1]\n",
    "\n",
    "query_list = list()\n",
    "\n",
    "ID_to_odb = dict()\n",
    "odb_to_OGset = dict()\n",
    "tax_dict = dict()\n",
    "OG_to_odbSet = dict()\n",
    "tax_dict = dict()\n",
    "uniprot_to_homologs = dict()\n",
    "\n",
    "# the query file is read in and a list of UniProt IDs is generated\n",
    "\n",
    "with open(query_file_name, 'r') as file_obj:\n",
    "    for line in file_obj:\n",
    "        line = line.rstrip()\n",
    "        query_list.append(line)\n",
    "\n",
    "# the genes.tab file is parsed into a two-level dictionary where the odb ID and tax ID are saved as the value for a uniprot ID key\n",
    "            \n",
    "with open('odb9v1_genes.tab', 'r') as file_obj:\n",
    "   for line in file_obj:\n",
    "        line_list = line.split()\n",
    "        ID_to_odb[line_list[3]] = {'odb' : line_list[0],'tax': line_list[1]}\n",
    "\n",
    "print('output of ID_to_odb:',ID_to_odb['O14802']) \n",
    "# this returns: output of ID_to_odb: {'odb': '9606:00232f', 'tax': '9606'}\n",
    "\n",
    "\n",
    "# the OG2genes.tab file is parse into a dictionary where a key of odb ID will return the OG\n",
    "\n",
    "        \n",
    "with open('odb9v1_OG2genes.tab', 'r') as file_obj:\n",
    "    for line in file_obj:\n",
    "        odbID = line.split()[1]\n",
    "        OG = line.split()[0]\n",
    "        if odbID in odb_to_OGset:\n",
    "            odb_to_OGset[odbID].append(OG)\n",
    "        else :\n",
    "            odb_to_OGset[odbID] = [OG]\n",
    "\n",
    "# this converts the list of OGs associated with each odb to a set\n",
    "            \n",
    "    for odbID in odb_to_OGset:\n",
    "        OG_list = odb_to_OGset[odbID]\n",
    "        OG_set = set(OG_list)\n",
    "        odb_to_OGset[odbID] = OG_set\n",
    "\n",
    "print('output of odb_to_OGset:',odb_to_OGset['9606:00232f'])\n",
    "# this returns: output of odb_to_OGset: {'EOG093700ID', 'EOG090B00IL', 'EOG090M00IS', 'EOG090700I0', 'EOG090O017L', 'EOG090A00I5', 'EOG090600QN', 'EOG0903023G', 'EOG091G0FNO'}\n",
    "\n",
    "# the OG_dict is reversed such that a key of an OG will return a list of odb IDs that are in that group\n",
    "        \n",
    "for odbID in odb_to_OGset:\n",
    "    OG_list = odb_to_OGset[odbID]\n",
    "\n",
    "    for ogID in OG_list:\n",
    "        if ogID in OG_to_odbSet:\n",
    "            OG_to_odbSet[ogID].add(odbID)\n",
    "        else :\n",
    "            OG_to_odbSet[ogID] = set()\n",
    "            OG_to_odbSet[ogID].add(odbID)\n",
    "\n",
    "print('output of OG_to_odb_list is:',OG_to_odbSet['EOG0903023G'])\n",
    "# this returns: output of OG_to_odb_list is: {'379532:0044ce', '9986:002416', '9601:000893', '34839:003337', '10020:0000a2', '1026970:004272', '37293:003aeb', '10042:004777', '9544:00113a', '29078:0013f4', '43179:0021f6', '54131:0007ca', '9598:00074b', '9568:00165b', '9478:0017aa', '60711:00233d', '30608:001525', '9595:004e3a', '30611:0020f0', '9483:000337', '10090:001347', '9978:00069f', '51337:002648', '37347:0033ce', '9555:003d10', '885580:004a23', '9531:0044ee', '10141:003b88', '9545:00068d', '9597:00030d', '9402:0024ed', '246437:0024f7', '10036:0018bc', '482537:001036', '10160:004262', '9606:00232f', '10116:001418', '61853:002648', '61622:00181b'} \n",
    "\n",
    "# a list of OGs associated with an odb will be collected, the list of odbs associated with those OGs are added to a set\n",
    "# so that a list of odbs is returned for a provided odb\n",
    "\n",
    "for uniprot_id in query_list:\n",
    "    homologs = set()\n",
    "    odb = ID_to_odb[uniprot_id][odb]\n",
    "    OGlist = odb_to_OGset[odb]\n",
    "\n",
    "    for OG in OGlist:\n",
    "        odbList = OG_to_odbSet[OG]\n",
    "\n",
    "        for related_odb in odbList:\n",
    "            homologs.add(related_odb)\n",
    "\n",
    "    uniprot_to_homologs[uniprot_id] = homologs\n",
    "\n",
    "print(uniprot_to_homologs)\n",
    "    \n",
    "# a lookup table is created from the parsed gene dict in which a odb ID as key will return tax ID\n",
    "\n",
    "with open('output.tab' , 'w') as outputFile:\n",
    "    for uniprot_ID in uniprot_to_homologs:\n",
    "        mapped_odbs = uniprot_to_homologs[uniprot_ID]\n",
    "\n",
    "        mapped_odb_str = ''\n",
    "        \n",
    "        for odb in mapped_odbs:\n",
    "            taxa = odb.split(':')[0]\n",
    "            mapped_odb_str += taxa+','\n",
    "\n",
    "        mapped_odb_str = mapped_odb_str.rstrip(',')\n",
    "            \n",
    "        outputFile.write(uniprot_ID+'\\t'+'['+mapped_odb_str+']'+'\\n')\n",
    "            "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
